src/platform_impl/web/web_sys/canvas.rs

```diff
@@ -158,97 +185,147 @@ impl Canvas {
+    {
+        self.on_cursor_move = Some(self.add_event(
+            "pointermove",
             move |event: PointerEvent| {
                 handler(
-                    event.pointer_id(),
-                    event::mouse_button(&event),
+                    event::mouse_position(&event),
                     event::mouse_modifiers(&event),
                 );
             },
+            false,
         ));
     }
 
-    pub fn on_cursor_move<F>(&mut self, mut handler: F)
+    pub fn on_fullscreen_change<F>(&mut self, mut handler: F)
     where
-        F: 'static + FnMut(i32, LogicalPosition, ModifiersState),
+        F: 'static + FnMut(),
     {
-        self.on_cursor_move = Some(self.add_event("pointermove", move |event: PointerEvent| {
-            handler(
-                event.pointer_id(),
-                event::mouse_position(&event),
-                event::mouse_modifiers(&event),
-            );
-        }));
+        self.on_fullscreen_change =
+            Some(self.add_event("fullscreenchange", move |_: Event| handler(), false));
     }
 
-    pub fn on_mouse_wheel<F>(&mut self, mut handler: F)
+    pub fn on_gamepad_connected<F>(&mut self, mut handler: F)
     where
-        F: 'static + FnMut(i32, MouseScrollDelta, ModifiersState),
+        F: 'static + FnMut(GamepadShared),
     {
-        self.on_mouse_wheel = Some(self.add_event("wheel", move |event: WheelEvent| {
-            if let Some(delta) = event::mouse_scroll_delta(&event) {
-                handler(0, delta, event::mouse_modifiers(&event));
-            }
-        }));
+        let m = self.gamepad_manager.clone();
+        self.on_gamepad_connected = Some(self.add_event(
+            "gamepadconnected",
+            move |event: GamepadEvent| {
+                let gamepad = event
+                    .gamepad()
+                    .expect("[gamepadconnected] expected gamepad");
+                let g = m.register(gamepad);
+                handler(g);
+            },
+            false,
+        ))
     }
 
-    pub fn on_fullscreen_change<F>(&mut self, mut handler: F)
+    pub fn on_gamepad_disconnected<F>(&mut self, mut handler: F)
     where
-        F: 'static + FnMut(),
+        F: 'static + FnMut(GamepadShared),
     {
-        self.on_fullscreen_change =
-            Some(self.add_event("fullscreenchange", move |_: Event| handler()));
+        let m = self.gamepad_manager.clone();
+        self.on_gamepad_disconnected = Some(self.add_event(
+            "gamepaddisconnected",
+            move |event: GamepadEvent| {
+                let gamepad = event
+                    .gamepad()
+                    .expect("[gamepaddisconnected] expected gamepad");
+                let g = m.register(gamepad);
+                handler(g);
+            },
+            false,
+        ))
     }
 
-    fn add_event<E, F>(&self, event_name: &str, mut handler: F) -> Closure<dyn FnMut(E)>
+    fn add_event<E, F>(
```

Comment:
I think using the same interface for Canvas and Window events will run into a duplicate events problem. If I have 'n' canvases, I shouldn't get 'n' window events.

Reply:
What do you suggest? Gamepad events are emitted in window.

Reply:
I created add_window_event

Reply:
This still doesn't solve the problem, because if I create 2 canvases I will get duplicate gamepad events. One option would be to create a new Window struct in the web backend that handles keeping track of gamepad events, separate from Canvas.

Reply:
Understood. I splitted canvas and window and attached a shared window to the window_target https://github.com/rust-windowing/winit/pull/1414/commits/7948cd0ffd3d9774c616c29a39f0fd29b01d45e4#diff-2a56338f5cb2485505fe3737f17c7753R26

Reply:
Splitted https://github.com/rust-windowing/winit/pull/1414/files#diff-9b74a17322bdee110961a29607d385e9

Reply:
👍